Support eth and consensus subprotocols

libby opened the pull request:
Enable the eth service to support multiple subprtocols, e.g.
"eth" AND "istanbul/100", instead of a single protocol, e.g. "eth" OR "istanbul/99".

When istanbul consensus was originally added, the eth service was
modified to override the eth subprotocol maintained by the eth service
to an quorum specific consensus protocol, e.g. "istanbul/99" meaning all
eth p2p messages would be sent over a consensus specific protocol in
addition to any consensus protocol message. In the case of istanbul,
the "eth" subprotocol would no long be included as a devp2p capability/
subprotocol for the node.

This commit changes that, and enables the eth service to return multiple
subprotocols, e.g. "eth", "istanbul/100".

"istanbul/100" has been added as a new istanbul subprotocol version, which supports
the multple subprotocols design. peers running older versions of istanbul will
continue to override the "eth" service and will communicate over a single subprotocol:

istanbul/99 <-> istanbul/99
istanbul/100, eth <-> istanbul/100, eth

To maintain backwards compatibility, when establishing a connection with a
new peer that does not support multiple protocols, but rather a "legacy"
protocol, e.g. "istanbul/64" "istanbul/99" "clique" (v2.6), the nodes will communicate
over the single "legacy" protocol and not "eth" + consensus subprotocol.

When running two subprotocols, "eth" + "consensus", the eth subprotocol
manages the lifecycle of the peer for the eth service, and the consensus
subprotocol uses this peer for its communication, registering a separate
protocol read writer on it.

maintains backwards compatibility with: istanbul, raft, clique.

libby committed:
add eth service support for multiple subprotocols.

Enable the eth service to support multiple subprotocols, e.g.
"eth" AND "istanbul/100", instead of a single protocol, e.g.
"eth" OR "istanbul/99".

When istanbul consensus was originally added, the eth service was
modified to override the eth subprotocol maintained by the eth service
to a quorum specific consensus protocol, e.g. "istanbul/99" meaning all
eth p2p messages would be sent over a consensus specific subprotocol in
addition to any consensus protocol message. In the case of istanbul,
the "eth" subprotocol would no long be included as a devp2p capability/
subprotocol for the node.

This commit changes that, and enables the eth service to return multiple
subprotocols, e.g. "eth" and "istanbul/100".

"istanbul/100" has been added as a new istanbul subprotocol version,
which supports the multple subprotocol design. peers running older
versions of istanbul will continue to override the "eth" service
and will communicate over a single subprotocol:

old node: istanbul/99 <-> istanbul/99
new node: istanbul/100, eth <-> istanbul/100, eth

To maintain backwards compatibility, when establishing a connection with a
new peer that does not support mutliple protocols, but rather a "legacy"
protocol, e.g. "istanbul/64" "istanbul/99" "clique" (v2.6), the nodes
will communicate over the single "legacy" protcol and not "eth" + consensus
subprotocol.

When running two subprotcols, "eth" + "istanbul" (consensus), the "eth"
subprotocol manages the lifecycle of the peer for the eth service, and
the consensus subprotocol uses this peer for its communication, registering
a separate protocol read writer on it.

To support this, a channel is added to a p2p peer (EthPeerRegistered), which
is used to communicate to the "consensus" / quorum subprotocols that the
eth peer is ready and can be retrieved.

When running a quorum specific subprotocol on the eth service, the eth
peer is started and managed by the "eth" subprotocol. The quorum specific
subprotocol uses this peer to send messages.

This channel is necessary, as every protocol registerd in a p2p peer
running map, the Run method is called inside a goroutine. When iterating
through the running map,  the order is not guaranteed and inside the
goroutines we cannot count on execution time of the methods. Thus, we
use the EthPeerRegistered channel to ensure that the eth protocol has
register the peer before attempting to look it up.

maintains backwards compatiblity with: istanbul, raft, clique.

libby committed:
notify ethPeerRegistered channel if peer disconnected.

libby committed:
eth: eth65 and multi eth service protocol.

When registering a peer check that the protocol is
"eth" and version >= eth65 and not a legacy quorum protocol to
determine if p.announceTransactions() can be called.

nmvalera commented on eth/handler.go:
```diff
@@ -391,6 +392,15 @@ func (pm *ProtocolManager) handle(p *peer) error {
+
+	// Quorum notify other subprotocols that the eth peer is ready, and has been added to the peerset.
+	select {
+	case p.EthPeerRegistered <- true:
```
I am not fan of notifying through a `chan bool` 

A more standard option is to use a `chan struct{}` and close it to broadcast notification.

In this case, since it is required to notify `true` or `false` I would instead use 2 channels 
- EthPeerReady
- EthPeerDisconnected

libby replied:
Changed `EthPeerRegistered` to buffered `chan struct{}{}` and added `EthPeerDisconnected` instead of using chan bool and sending `true` `false`.

nmvalera commented on eth/quorum_protocol.go:
```diff
@@ -0,0 +1,216 @@
+
+// istanbul/64, istanbul/99, clique/63, clique/64 all override the "eth" subprotocol.
+func isLegacyProtocol(name string, version uint) bool {
+	legacyProtocols := map[string][]uint{"istanbul": {64, 99}, "clique": {63, 64}}
```
Can we declare `legacyProtocols` as a global variable?

libby replied:
We can, but it is only used in the `isLegacyProtocol` method atm.

ricardolyn commented on eth/backend.go:
```diff
@@ -202,6 +211,7 @@ func New(ctx *node.ServiceContext, config *Config) (*Ethereum, error) {
 		dbVer = fmt.Sprintf("%d", *bcVersion)
 	}
 	log.Info("Initialising Ethereum protocol", "name", protocolName, "versions", ProtocolVersions, "network", config.NetworkId, "dbversion", dbVer)
+	log.Info("Initialising Quorum consensus protocol", "name", quorumConsensusProtocolName, "versions", quorumConsensusProtocolVersions, "network", config.NetworkId, "dbversion", dbVer)
```
should we check `if chainConfig.IsQuorum` before the log? or show a different log

libby replied:
sure, added a `if chainConfig.IsQuorum` check.

ricardolyn commented on eth/backend.go:
```diff
@@ -579,6 +600,15 @@ func (s *Ethereum) Protocols() []p2p.Protocol {
+
+	// /Quorum
+	// add additional quorum consensus protocol if set and if not set to "eth", e.g. istanbul
+	if quorumConsensusProtocolName != "" && quorumConsensusProtocolName != protocolName {
```
Q: in which situation `quorumConsensusProtocolName` could get the value `eth`?

libby replied:
In the case where the consensus engine return `eth` as the consensus protocol, e.g. `clique`.

ricardolyn commented on eth/quorum_protocol.go:
```diff
@@ -0,0 +1,216 @@
+// The Run method starts the protocol and is called by the p2p server. The quorum consensus subprotocol,
+// leverages the peer created and managed by the "eth" subprotocol.
+// The quorum consensus protocol requires that the "eth" protocol is running as well.
+func (pm *ProtocolManager) makeQuorumConsensusProtocol(ProtoName string, version uint, length uint64) p2p.Protocol {
```
do we have unit tests for this new code or other tests that are covering this?

ricardolyn reviewed (commented):
is there a plan to add Acceptance Tests for this case?

nmvalera commented on eth/backend.go:
```diff
@@ -186,10 +186,19 @@ func New(ctx *node.ServiceContext, config *Config) (*Ethereum, error) {
-		protocolName = quorumProtocol.Name
-		ProtocolVersions = quorumProtocol.Versions
+		// set the quorum specific consensus devp2p subprotocol, eth subprotocol remains set to protocolName as in upstream geth.
+		quorumConsensusProtocolName = quorumProtocol.Name
```
Is it necessary to use `quorumConsensusProtocolName`, `quorumConsensusProtocolVersions`, etc. 
global variable?

libby replied:
We are mirroring the way the eth `protocolName` `ProtocolVersions` is set for consistency:
eth/backend.go
```
func (s *Ethereum) Protocols() []p2p.Protocol {
	protos := make([]p2p.Protocol, len(ProtocolVersions))
	for i, vsn := range ProtocolVersions {
```	
```
func (s *Ethereum) quorumConsensusProtocols() []p2p.Protocol {
	protos := make([]p2p.Protocol, len(quorumConsensusProtocolVersions))
	for i, vsn := range quorumConsensusProtocolVersions {
```

nmvalera replied:
Ok I get the point.

Still I see a difference

```go
func (s *Ethereum) Protocols() []p2p.Protocol {
	protos := make([]p2p.Protocol, len(ProtocolVersions))
	for i, vsn := range ProtocolVersions {
```

uses `ProtocolVersions` global variable which is constant while

```go
func (s *Ethereum) quorumConsensusProtocols() []p2p.Protocol {
	protos := make([]p2p.Protocol, len(quorumConsensusProtocolVersions))
	for i, vsn := range quorumConsensusProtocolVersions {
```

uses `quorumConsensusProtocolVersions` global variable which is not constant. 

`quorumConsensusProtocolVersions` is indirectly set by `New(...) (*Ethereum, error)` which I think is not a good practice and maybe error-prone (e.g it creates coupling between `Ethereum` structs through a global variable which is not constant).

I think it would be better to have a `quorumConsensusProtocolVersions` attribute on the `Ethereum` struct that is set on `New(...)` then `quorumConsensusProtocols()` could re-use the attribute or something like this.

What do you think?

libby replied:
Before this change, the `ProtocolVersions` and `ProtocolName` vars would be overridden here to set them to the  quorum specific subprototol obtained from the consensus engine.  In upstream (same geth version), they would not be overridden. This change leaves `ProtocolName` `ProtocolVersions` always set to the `eth` subprotocol. 
With this initial dual subprotocol approach, keeping the changes as minimally invasive to the existing code as possible was the approach. In the (near) future, this will be refactored more (consensus may be a service with its own protocol, when we get up to date with the current geth, eth service will be refactored to support more than one subprotocol, e.g. `snap`, `eth`, `quorumconsensus`. To the client, they won't see a difference once they are running the dual subprotocols, but the Quorum code can refactored to support a different subprotocol design.

nmvalera replied:
Ok

libby committed:
eth,p2p: make EthPeerRegistered chan struct{}.

make EthPeerRegistered a buffered chan struct{} instead of a chan bool,
add EthPeerDisconnected chan to listen for peer disconnect.

nmvalera reviewed (approved):
LGTM
